### PR TITLE
Fixed the ConfigParser failure with paths with percent

### PR DESCRIPTION
**Pull request by ssbarnea**

### CODE REVIEW

File: numpy/distutils/system_info.py [hunk 1]
```diff
@@ -128,9 +128,13 @@
...
 from functools import reduce
 if sys.version_info[0] < 3:
-    from ConfigParser import NoOptionError, ConfigParser
+    from ConfigParser import NoOptionError
+    from ConfigParser import RawConfigParser as ConfigParser
 else:
-    from configparser import NoOptionError, ConfigParser
+    from configparser import NoOptionError
+    from configparser import RawConfigParser as ConfigParser
+    # It seems that some people are importing ConfigParser from here so is good to keep its class name
```
Comment (charris) [inline numpy/distutils/system_info.py]: Lines < 80 characters in length.

**Comment by charris** [comment]
Please squash your commits into one and reformat your commit messages with all lines <= 72 characters. See  `doc/source/dev/gitwash/development_workflow.rst` for commit message formats.

**Comment by charris** [comment]
Also, make the commit against master rather than 1.11. All merges start in master.

**Comment by charris** [comment]
Also make a new branch for the work, don't use an existing branch.

**Comment by pv** [comment]
Why does it actually fail? Presumably, the data is not read from a configuration file?

**Comment by ssbarnea** [comment]
The problem is quite complex an not very easy to replicate because it happens only if you current PATH contains a percent in it and when you are using virtual environment. While the second is quite common the first one will happen for people using git flow when they will enable CI builds on their feature br
anches (which have a slash in their name).

Here is some code that will help you replicate the problem on any Linux machine:
```
#!/bin/bash
set -ex

WORKDIR="feature%2Fcool"
mkdir -p ${WORKDIR}
cd ${WORKDIR}

PYENV="env"
mkdir -p $PYENV

if [[ ! -d "${PYENV}/bin" ]]; then
    virtualenv --no-site-packages ${PYENV}
fi

source ${PYENV}/bin/activate

pip install --no-binary :all: numpy
echo "done"
```
Now, if you try to install numpy from my fork, the one that switched the use of ConfigParser with RawConfigParser everything will be working fine.

NumPy is not the only package that is affected by this issue and it would be important to note that this happens only with Python 3.

**Comment by ssbarnea** [comment]
If you want to understand better the problem of the 3 ConfigParser implementations you can check this https://travis-ci.org/ssbarnea/test-configparser/jobs/126210802 and the code from https://github.com/ssbarnea/test-configparser/blob/master/tests/test-configparser.py#L21

RawConfigParser is the only class that is able to load `value=something%2Felse` from a config file on both Python 2 and Python 3. SafeConfigParser fails all the time, and ConfigParser has different behaviour on Python 2 (works) than Python 3 (fails).

**Comment by ssbarnea** [comment]
@pv please have a look at this gist if you want to see how this bug behaves in the real life https://gist.github.com/ssbarnea/b373062dd45de92735c7482b2735c5fb 

I have no idea what is this file and who generates it but I do know that config files are not supposed to have percent characters escaped in them anyway, so why to ask for something that is not normal anyway?

Even Python documentation states the RFC 822 which does NOT contain any escaping of percent sign.

The only reference to percent is made is as an additional feature named interpolation which is added by the SafeConfigParser class (unfortunate name). And because Python 3 decided to use SafeConfigParser as the default ConfigParser the hells gates are not open because code that worked file on Python 3 to load values containing percent in them will fail unless the code is modified to use the RawConfigParser - the one that does not do any interpolation. 

Opposed to the assumption the Safe is better, RawConfigParser is as safe as the other other one, is just that it does not have the interpolation anti-feature.

**Comment by pv** [comment]
Do you know what component fails to escape the percent sign? Presumably,
there could (?) be site.cfg files that use the substitution feature.

**Comment by ssbarnea** [comment]
The problem is that the temporary /tmp/easy_install... are removed after the error and I wasn't able to find a way to make them stay on disk in order to look inside them and see why they fail.

Anyway one thing is clear: any usage of `ConfigParser` class in py2+py3 software is clearly a bug because this class behaves different across those two versions of Python. If the code was supposed to use interpolation it should have used SafeConfigParser, if not it should have used RawConfigParser. The one that does not contain Safe or Raw prefix is clearly a bug, as it has unexpected behaviour.

**Comment by ssbarnea** [comment]
I even raise the bug against Python but I doubt we can wait for this to be fixed http://bugs.python.org/issue26872

**Comment by njsmith** [comment]
I guess what's unclear to me is where this .cfg file is coming from. Do you
know what's generating it?

On Wed, Apr 27, 2016 at 2:12 PM, Sorin Sbarnea <notifications@github.com>
wrote:

> I even raise the bug against Python but I doubt we can wait for this to be
> fixed http://bugs.python.org/issue26872
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/7576#issuecomment-215230817>
>



-- 
Nathaniel J. Smith -- https://vorpus.org <http://vorpus.org>

**Comment by ssbarnea** [comment]
The installation process is quite complex and I am not even sure that this file exists on disc, it may be possible that is using a file stream object to load that.

Anyway here is the additional information, I was able to get:

This is the line that generates the issue, `distutils/system_info.py` inside `get_paths`:
```
            dirs = self.cp.get(section, key).split(os.pathsep)
```
It is called with these parameters:
```
key: 'library_dirs'
section: 'mkl'
self: <numpy.distutils.system_info.blas_mkl_info object at 0x...>
```
And it makes sense because the `library_dirs` do contain a percent in them.

Again, the bug happens only with Python 3 because Python 3 decided to map the ConfigParser lass to SafeConfigParser instead of RawConfigParser.

Nobody wants interpolation here and I would say that in most cases people should not use it as it a security risk, causes problems like this and more important: interpolation is not compliant with the RFC. RawConfigParser is compliant so just start using it instead of ConfigParser.

Now, there is a small catch regarding how we should fix this bug: just replacing ConfigParser occurrences with RawConfigParser would work for numpy project itself but it seems that there are few modules (like numexpr) which decided to import the ConfigParser from numpy/distutils which means that changing the import in our numpy will break them.

For this reason I decided to keep the interface the same and to import the `RawConfigParser as ConfigParser`, avoiding breaking the code of those that decided to re-use our import from other modules.

Again, if someone wonders about "what if someone was using interpolations..." - nope, nobody was using interpolations because these would have worked only in Python 3 and not in Python 2.

Based on the evidence that I was able to gather in the last days, now I can say that if a py2+py3 project is using the ConfigParser, they are clearly doing something wrong. You either want to use RawConfigParser or SafeConfigParser, but never the ConfigParser.

**Comment by ssbarnea** [comment]
@pv @charris Based on the requests received earlier, I am closing this PR in favour of https://github.com/numpy/numpy/pull/7584 which is made from a fix branch towards master.

**Pull request closed without merging**